### PR TITLE
Fix for Sending Notification to Slack Channel on List of PRs in LSP4IJ

### PR DESCRIPTION
**Pull request by vaisakhkannan**
Fixes https://github.com/OpenLiberty/liberty-tools-intellij/issues/803
Added a new YAML file, `pr-slack-notification.yaml,` for sending Slack notifications based on cron job runs.

Fixes https://github.com/OpenLiberty/liberty-tools-intellij/issues/810

At the end of `pr-slack-notification.yaml`, I have added one more job called `slack-notification`, which is used to trigger Slack Workflow Builder as a fix for the issue https://github.com/OpenLiberty/liberty-tools-intellij/issues/810

### CODE REVIEW

**Comment by vaisakhkannan** [comment]
When I checked during testing in the cron job runs, I saw that fetching the Closed PR list was correct, but it was failing to send the payload. When I tested it locally, it was working. So I made a change to send it as a single payload that can be used for sending both open and closed PR lists in a single step in the workflow file. I hope it will work in today's cron jobs.

File: .github/workflows/pr-slack-notification.yaml [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+            cat cache/notified_prs.json
+          else
+            echo "Cache not restored or file does not exist.."
+          fi
+
+      - name: Fetch all opened pull request details using condition
+        id: fetch_all_pull_requests_using_condition
+        run: |
+          pr_infos=$(curl -s -H "Authorization: token ${{ secrets.GITHUB_TOKEN }}" \
+          "${{ env.API_URL }}?state=open&sort=created&direction=desc&per_page=100")
```
Comment (aparnamichael) [inline .github/workflows/pr-slack-notification.yaml:64]: sort=created, why are we sorting like this?
In lsp4ij PR's list, I am not seeing any value 'created' for sort parameter.

File: .github/workflows/pr-slack-notification.yaml [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+        with:
+          path: cache
+          key: ${{ env.CACHE_KEY }}
+
+      - name: Ensure cache directory exists and check if cache restored
+        run: |
+          mkdir -p cache
+          if [ -f cache/notified_prs.json ]; then
+            echo "Cache restored successfully."
+            cat cache/notified_prs.json
```
Comment (aparnamichael) [inline .github/workflows/pr-slack-notification.yaml:55]: Do we need this cat command here?

File: .github/workflows/pr-slack-notification.yaml [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+            pr_user=$(_jq '.user.login')
+            pr_url=$(_jq '.html_url')
+            pr_draft=$(_jq '.draft')
+            pr_created_at=$(_jq '.created_at')
+            pr_updated_at=$(_jq '.updated_at')
+
+            pr_data=$(jq -n --arg number "$pr_number" --arg updated_at "$pr_updated_at" '{number: $number, updated_at: $updated_at}')
+            new_notified_prs=$(echo "$new_notified_prs" | jq --argjson pr_data "$pr_data" '. += [$pr_data]')
+
+            echo "New Notified PR List------ $new_notified_prs"
```
Comment (aparnamichael) [inline .github/workflows/pr-slack-notification.yaml:93]: Please remove unwanted echo statements.

File: .github/workflows/pr-slack-notification.yaml [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+
+            # Check if the PR is new or updated
+            previous_pr=$(echo "$previous_prs" | jq --arg number "$pr_number" '.[] | select(.number == $number)')
+            echo "Checking PR is new or not --- $previous_pr"
+            if [ -z "$previous_pr" ] || [ "$(echo "$previous_pr" | jq -r '.updated_at')" != "$pr_updated_at" ]; then
+              draft_status=""
+              echo "------Draft value ---- $pr_draft"
```
Comment (aparnamichael) [inline .github/workflows/pr-slack-notification.yaml:100]: Please remove unwanted echo statements.

File: .github/workflows/pr-slack-notification.yaml [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+          if [ -f cache/last_run_timestamp.txt ]; then
+            last_run_timestamp=$(cat cache/last_run_timestamp.txt)
+          else
+            last_run_timestamp=$(date -u -d "-24 hours" +%Y-%m-%dT%H:%M:%SZ)
+          fi
+
+          echo "Last run timestamp: $last_run_timestamp"
+
+          closed_prs=$(curl -s -H "Authorization: token ${{ secrets.GITHUB_TOKEN }}" \
+          "${{ env.API_URL }}?state=closed&sort=updated&direction=desc&per_page=100")
```
Comment (aparnamichael) [inline .github/workflows/pr-slack-notification.yaml:133]: why are we using this parameter "sort=updated"?

File: .github/workflows/pr-slack-notification.yaml [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+
+          closed_pr_list=""
+
+          for pr_info in $(echo "$closed_prs" | jq -r '.[] | @base64'); do
```
Comment (aparnamichael) [inline .github/workflows/pr-slack-notification.yaml:137]: These lines of code are present in step "Fetch all opened pull request details using condition" and here. Is there is any way to avoid this code duplication?

File: .github/workflows/pr-slack-notification.yaml [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+          payload=$(cat <<-EOF
+          {
+            "blocks": [
+              $(IFS=,; echo "${payload_blocks[*]}")
+            ]
+          }
+          EOF
+          )
+
+          # Debugging: Print payload to verify its format
```
Comment (aparnamichael) [inline .github/workflows/pr-slack-notification.yaml:229]: Remove print statements used for debugging.

File: .github/workflows/pr-slack-notification.yaml (hunk 2 above)
Comment (vaisakhkannan) [reply .github/workflows/pr-slack-notification.yaml:55]: This is just for printing to ensure that the cache was fetched correctly, I hope this will help in the future to check the fetched data correctly. If required, we can delete it.

Screenshot 2024-06-25 at 2 17 40 PM

File: .github/workflows/pr-slack-notification.yaml (hunk 3 above)
Comment (vaisakhkannan) [reply .github/workflows/pr-slack-notification.yaml:93]: If we comment out this echo statement, it will be difficult to understand the workflow logs in the future. We won't be able to debug this process. The only way to analyze the workflow is by using echo statements. If we remove them now, we might need to modify the code later when checking the process (including necessary changes like secrets).We may want to rerun the workflow which will goes repeated unwanted noice in slack channel. This could disrupt the current working code. This makes it challenging for other developers, especially those who have not worked on this previously. Clear workflow logs help easily identify issues.

File: .github/workflows/pr-slack-notification.yaml (hunk 6 above)
Comment (vaisakhkannan) [reply .github/workflows/pr-slack-notification.yaml:137]: It's challenging to avoid code duplication by creating a reusable step or script that handles the common functionality between 'Fetch closed pull requests since the last run' and 'Fetch all opened pull request details using conditions' because both steps involve different logics. Additionally, we may require additional conditional checks, which complicates efforts to prevent code duplication.

File: .github/workflows/pr-slack-notification.yaml (hunk 7 above)
Comment (vaisakhkannan) [reply .github/workflows/pr-slack-notification.yaml]: Yes I removed it.

File: .github/workflows/pr-slack-notification.yaml (hunk 5 above)
Comment (vaisakhkannan) [reply .github/workflows/pr-slack-notification.yaml]: Yes I modified the code.

File: .github/workflows/pr-slack-notification.yaml (hunk 4 above)
Comment (vaisakhkannan) [reply .github/workflows/pr-slack-notification.yaml]: Yes I modified it

File: .github/workflows/pr-slack-notification.yaml (hunk 1 above)
Comment (vaisakhkannan) [reply .github/workflows/pr-slack-notification.yaml]: Yes I modified it.

File: .github/workflows/pr-slack-notification.yaml (hunk 3 above)
Comment (aparnamichael) [reply .github/workflows/pr-slack-notification.yaml:93]: @vaisakhkannan  Yes, logs help us understand the results at each step, but adding more logs can affect the performance especially inside the iterative statements like for/while loops and it will make. Also it will reduce the readability . For example, the screenshot you have attached shows logs for each PR detail with "Checking PR is new or not," followed by printing the list again with "New Notified PR List."

Instead, you could simply log the list of PRs retrieved by the API and the list of new or updated PRs, right?

File: .github/workflows/pr-slack-notification.yaml (hunk 5 above)
Comment (vaisakhkannan) [reply .github/workflows/pr-slack-notification.yaml]: The GitHub API expects certain values for the state parameter, such as `open, closed, or all`, but it does not support combining states like `(closed || open)` directly.

If we want to fetch both open and closed pull requests, we can use `state=all` to include all pull requests regardless of their state. However, this approach has drawbacks. It includes all PRs (both open and closed) in the comparison with the `cached PR list`, which increases the time and computational requirements. Additionally, using `state=all` prevents us from specifying a limit like `per_page=100`, as it will fetch all PRs in the repository (currently around 384 PRs in lisp4ij).

Screenshot 2024-06-26 at 10 23 57 AM

File: .github/workflows/pr-slack-notification.yaml [hunk 8]
```diff
@@ -0,0 +1,297 @@
...
+          
+          if [ -n "$mostRecentCacheKey" ]; then
+            echo "CACHE_KEY=$mostRecentCacheKey" >> $GITHUB_ENV
+          else
+            echo "No recent cache key found, generating a new one...."
+            echo "CACHE_KEY=${{ runner.os }}-pr-cache-${{ github.run_number }}" >> $GITHUB_ENV
+          fi
+          
+          echo "Done"
+          # 'secrets.GITHUB_TOKEN' is automatically provided by GitHub Actions for each workflow run. You don't need to manually create or manage this token.
```
Comment (mrglavas) [inline .github/workflows/pr-slack-notification.yaml:41]: Is this comment in the right place? It doesn't seem related to the line of code immediately above it.

File: .github/workflows/pr-slack-notification.yaml [hunk 9]
```diff
@@ -0,0 +1,297 @@
...
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v2
+
+      - name: Fetch recent cache
+        run: |
+          gh extension install actions/gh-actions-cache
+
+          # Fetching list of cache keys........
+          cacheKeys=$(gh actions-cache list -R $REPO -B $BRANCH -L 100 | cut -f 1 )
```
Comment (mrglavas) [inline .github/workflows/pr-slack-notification.yaml:28]: I guess this is limiting the list of cache keys returned to 100. Are we likely to hit this limit? What would happen if we do?

File: .github/workflows/pr-slack-notification.yaml [hunk 10]
```diff
@@ -0,0 +1,297 @@
...
+            cat cache/notified_prs.json
+          else
+            echo "Cache not restored or file does not exist.."
+          fi
+
+      - name: Fetch all opened pull request details using condition
+        id: fetch_all_pull_requests_using_condition
+        run: |
+          pr_infos=$(curl -s -H "Authorization: token ${{ secrets.GITHUB_TOKEN }}" \
+          "${{ env.API_URL }}?state=open&direction=desc&per_page=100")
```
Comment (mrglavas) [inline .github/workflows/pr-slack-notification.yaml:64]: If I'm reading this correctly it looks like this is limited to fetching 100 open PRs (per page). Hopefully there are never that many open PRs but I'm curious if this code would work if there's more than 100.

File: .github/workflows/pr-slack-notification.yaml [hunk 11]
```diff
@@ -0,0 +1,297 @@
...
+          if [ -f cache/last_run_timestamp.txt ]; then
+            last_run_timestamp=$(cat cache/last_run_timestamp.txt)
+          else
+            last_run_timestamp=$(date -u -d "-24 hours" +%Y-%m-%dT%H:%M:%SZ)
+          fi
+
+          echo "Last run timestamp: $last_run_timestamp"
+
+          closed_prs=$(curl -s -H "Authorization: token ${{ secrets.GITHUB_TOKEN }}" \
+          "${{ env.API_URL }}?state=closed&direction=desc&per_page=100")
```
Comment (mrglavas) [inline .github/workflows/pr-slack-notification.yaml:131]: Similar question about the limit on the number closed PRs. Would this work if there's more than 100?

File: .github/workflows/pr-slack-notification.yaml (hunk 3 above)
Comment (TrevCraw) [reply .github/workflows/pr-slack-notification.yaml:93]: FYI for the scenario of this workflow, performance is not very important. Readability is highly important, though.

File: .github/workflows/pr-slack-notification.yaml (hunk 5 above)
Comment (TrevCraw) [reply .github/workflows/pr-slack-notification.yaml]: @vaisakhkannan I noticed you removed `sort=updated` without explaining what it was for. Was there a reason you had that parameter initially?

Comment (vaisakhkannan) [reply .github/workflows/pr-slack-notification.yaml]: @TrevCraw Initially, I thought it would be better to sort the closed PRs based on `sort=updated`. However, then I had realized that if we ran daily cron jobs, there was no need to sort, and the list may be within the 100 limit. If the number of closed PRs exceeds this limit, then we will definitely need `sort=updated`. So now I think based on Michael's comment aswell, I have added back to using `sort=updated`. In the future, we expect to have more than 100 closed PRs."

Screenshot 2024-06-27 at 4 28 11 PM

File: .github/workflows/pr-slack-notification.yaml (hunk 1 above)
Comment (vaisakhkannan) [reply .github/workflows/pr-slack-notification.yaml]: By default is `sort=created`. So I removed this. If require we can add it back.
 https://docs.github.com/en/rest/pulls/pulls?apiVersion=2022-11-28

File: .github/workflows/pr-slack-notification.yaml (hunk 3 above)
Comment (vaisakhkannan) [reply .github/workflows/pr-slack-notification.yaml]: @aparnamichael, @TrevCraw , I modified the code to list currently Opened PRs and List of PRs that need to be sent to Slack after that current run

Screenshot 2024-06-27 at 4 57 56 PM

File: .github/workflows/pr-slack-notification.yaml (hunk 6 above)
Comment (vaisakhkannan) [reply .github/workflows/pr-slack-notification.yaml]: @aparnamichael , I simplified the code for the step `'Fetch closed pull requests since the last run'`, and I attempted the same approach to simplify the code for the step `'Fetch all opened pull request details using conditions'` as well. I avoided using a for loop and instead utilized jq with select queries to filter out values. However, when checking the conditions we implemented for listing PRs (specifically, `omitting PRs already listed in Slack or those that have been updated or are new`), it did not succeed. If necessary, I will investigate further.

File: .github/workflows/pr-slack-notification.yaml (hunk 9 above)
Comment (vaisakhkannan) [reply .github/workflows/pr-slack-notification.yaml:28]: @mrglavas , Allowed values are within this limit `1-100` , Only the latest 100 cache keys will be listed. Any older cache keys beyond this limit will not be included in` $cacheKeys`. I am hoping that last cron job key will not be beyond the limit of 100.

File: .github/workflows/pr-slack-notification.yaml (hunk 10 above)
Comment (vaisakhkannan) [reply .github/workflows/pr-slack-notification.yaml:63]: @mrglavas , During my investigation , I found that "The number of results per page (max 100)". We  [pagination](https://docs.github.com/en/rest/using-the-rest-api/using-pagination-in-the-rest-api?apiVersion=2022-11-28) , if we need to increase the limit.

With this code, PR list will be limited to 100 and we can sort PRs based on creation date by "sort=created" .

File: .github/workflows/pr-slack-notification.yaml (hunk 9 above)
Comment (mrglavas) [reply .github/workflows/pr-slack-notification.yaml:28]: @vaisakhkannan It sounds unlikely that we would hit the limit but it would be good to document that the workflow has this limitation.

Comment (vaisakhkannan) [reply .github/workflows/pr-slack-notification.yaml:29]: @mrglavas, I added the comment in the code

File: .github/workflows/pr-slack-notification.yaml [hunk 12]
```diff
@@ -0,0 +1,287 @@
...
+            gh actions-cache delete $mostRecentCacheKey -R $REPO -B $BRANCH --confirm
+          else
+            echo "No cache keys found."
+          fi
+
+          echo "Done"
+
+    # 'secrets.GITHUB_TOKEN' is automatically provided by GitHub Actions for each workflow run. We don't need to manually create or manage this token.
+    # The below way, the workflow is granting write access to the actions scope of the GITHUB_TOKEN.
+    # If the below specific way is not successful to delete the cache key, then the repository owner will need to create a Personal Access Token (PAT) with write permissions, add it as a secret, and specify the name in GH_TOKEN.
```
Comment (TrevCraw) [inline .github/workflows/pr-slack-notification.yaml:262]: I am confused by the location of these comments

File: .github/workflows/pr-slack-notification.yaml (hunk 6 above)
Comment (TrevCraw) [reply .github/workflows/pr-slack-notification.yaml]: @vaisakhkannan Thank you for simplifying the code where able. If it is becoming too complicated or too complex to simplify the code elsewhere, there is no need to do so.

File: .github/workflows/pr-slack-notification.yaml (hunk 12 above)
Comment (vaisakhkannan) [reply .github/workflows/pr-slack-notification.yaml:262]: @TrevCraw , I added the comment for the below code.
```
permissions:
      actions: write
```

I mentioned this in the box note that this code is responsible for deleting the `cache key`. If we do not specify it this way, the deletion of the cache key cannot happen, and we will need a PAT, which needs to be added as a secret for deletion.

Screenshot 2024-06-29 at 11 03 15 AM

For reference : https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#defining-access-for-the-github_token-scopes-1

If require, we can modify the comment.

File: .github/workflows/pr-slack-notification.yaml [hunk 13]
```diff
@@ -0,0 +1,287 @@
...
+          else
+            echo "Cache not restored or file does not exist.."
+          fi
+
+      - name: Fetch all opened pull request details using condition
+        id: fetch_all_pull_requests_using_condition
+        run: |
+          # The number of results per page is limited to a maximum of 100. We need "pagination", if we require more than this limit, but for open PRs, it will be within the limit 100.
+          pr_infos=$(curl -s -H "Authorization: token ${{ env.GH_TOKEN }}" \
+          "${{ env.API_URL }}?state=open&direction=desc&per_page=100")
```
Comment (anusreelakshmi934) [inline .github/workflows/pr-slack-notification.yaml:65]: Is it necessary to specify state=open&direction=desc since the default values are state=open and direction=desc?

Comment (vaisakhkannan) [reply .github/workflows/pr-slack-notification.yaml:65]: I believe that, its good to keep `state=open@direction=desc` , because the default may change in future, so it will break the code when we try to fetch Open PRs

File: .github/workflows/pr-slack-notification.yaml [hunk 14]
```diff
@@ -0,0 +1,287 @@
...
+          
+          echo "List of PRs that need to be sent to Slack now:"
+          echo "$pr_list"
+
+          # Save current PR data for future comparison
+          echo "$new_notified_prs" > cache/notified_prs.json
+
+          if [ "$notify" = true ]; then
+            echo -e "$pr_list" > pr_list.txt
+            echo "::set-output name=notify::true"
```
Comment (anusreelakshmi934) [inline .github/workflows/pr-slack-notification.yaml:118]: There is a warning in the workflow file - `The `set-output` command is deprecated and will be disabled soon. Please upgrade to using Environment Files.`. So it is better to change set-output to the GITHUB_OUTPUT environment files. Ref - https://github.blog/changelog/2022-10-11-github-actions-deprecating-save-state-and-set-output-commands/

**Comment by aparnamichael** [comment]
@vaisakhkannan Can you please share the link of one workflow run?

File: .github/workflows/pr-slack-notification.yaml [hunk 15]
```diff
@@ -0,0 +1,287 @@
...
+            pr_number=$(_jq '.number')
+            pr_title=$(_jq '.title')
+            pr_user=$(_jq '.user.login')
+            pr_url=$(_jq '.html_url')
+            pr_draft=$(_jq '.draft')
+            pr_created_at=$(_jq '.created_at')
+            pr_updated_at=$(_jq '.updated_at')
+
+            pr_data=$(jq -n --arg number "$pr_number" --arg updated_at "$pr_updated_at" '{number: $number, updated_at: $updated_at}')
+            new_notified_prs=$(echo "$new_notified_prs" | jq --argjson pr_data "$pr_data" '. += [$pr_data]')
```
Comment (aparnamichael) [inline .github/workflows/pr-slack-notification.yaml:95]: What do we store in 'pr_data' variable? 
Why can't we directly parse and assign value to 'new_notified_prs' variable?

File: .github/workflows/pr-slack-notification.yaml [hunk 16]
```diff
@@ -0,0 +1,287 @@
...
+
+            # Check if the PR is new or updated
+            previous_pr=$(echo "$previous_prs" | jq --arg number "$pr_number" '.[] | select(.number == $number)')
+            if [ -z "$previous_pr" ] || [ "$(echo "$previous_pr" | jq -r '.updated_at')" != "$pr_updated_at" ]; then
```
Comment (aparnamichael) [inline .github/workflows/pr-slack-notification.yaml:99]: Why are we checking these conditions for getting draft status? In can see from the api itself we are getting the draft status.

File: .github/workflows/pr-slack-notification.yaml (hunk 12 above)
Comment (TrevCraw) [reply .github/workflows/pr-slack-notification.yaml:262]: @vaisakhkannan Thank you for the explanation. We should make it clear that the write permission enables the cache key to be deleted. I would recommend the below modifications:
```suggestion
    # Below, the workflow is granting "write" access to the actions scope of the GITHUB_TOKEN. This enables the workflow to be able to delete cache keys.
    # If the "write" permission does not enable the ability to delete the cache keys, then the repository owner will need to create a Personal Access Token (PAT) with write permissions, add it as a secret, and then specify the name in GH_TOKEN.
```

**Comment by vaisakhkannan** [comment]
@aparnamichael,  The link of workflow run --> https://github.com/vaisakhkannan/liberty-tools-intellij/actions/runs/9780833228/job/27003430709

File: .github/workflows/pr-slack-notification.yaml (hunk 15 above)
Comment (vaisakhkannan) [reply .github/workflows/pr-slack-notification.yaml:95]: @aparnamichael , Yes, when I tried to directly parse and assign the value to 'new_notified_prs', it made it difficult to read.

File: .github/workflows/pr-slack-notification.yaml (hunk 14 above)
Comment (vaisakhkannan) [reply .github/workflows/pr-slack-notification.yaml]: @anusreelakshmi934 , I modified the code to use `$GITHUB_ENV` , because this is using for pass values from one step to another step within the same job. Now I am not getting the warning.

Example of recent run: https://github.com/vaisakhkannan/liberty-tools-intellij/actions/runs/9793149679/job/27040487378

**Review by aparnamichael — DISMISSED** [review]
Looks good!

**Review by anusreelakshmi934 — APPROVED** [review]
Looks Good. Thanks

File: .github/workflows/pr-slack-notification.yaml [hunk 17]
```diff
@@ -0,0 +1,289 @@
...
+            gh actions-cache delete $mostRecentCacheKey -R $REPO -B $BRANCH --confirm
+          else
+            echo "No cache keys found."
+          fi
+
+          echo "Done"
+
+    # 'secrets.GITHUB_TOKEN' is automatically provided by GitHub Actions for each workflow run. We don't need to manually create or manage this token.
+    # The below way, the workflow is granting write access to the actions scope of the GITHUB_TOKEN.
+    # Below, the workflow is granting "write" access to the actions scope of the GITHUB_TOKEN. This enables the workflow to be able to delete cache keys.
```
Comment (TrevCraw) [inline .github/workflows/pr-slack-notification.yaml:263]: Line 263 was intended to replace 262

**Review by TrevCraw — CHANGES_REQUESTED** [review]

File: .github/workflows/pr-slack-notification.yaml (hunk 17 above)
Comment (vaisakhkannan) [reply .github/workflows/pr-slack-notification.yaml:262]: @TrevCraw , I removed the line 262.

**Review by aparnamichael — APPROVED** [review]

**Review by mrglavas — APPROVED** [review]

**Review by TrevCraw — APPROVED** [review]
